qiskit_experiments/test/mock_iq_backend.py

```diff
@@ -193,60 +204,146 @@ def _default_options(self):
+                memory.append(shot_memory)
+                shot_num += 1
 
         return memory
 
-    @abstractmethod
-    def _compute_probability(self, circuit: QuantumCircuit) -> float:
-        """Compute the probability used in the binomial distribution creating the IQ shot.
-
-        An abstract method that subclasses will implement to create a probability of
-        being in the excited state based on the received quantum circuit.
-
+    def _generate_data(
+        self, prob_dict: Dict[str, float], circuit: QuantumCircuit
+    ) -> Dict[str, Any]:
+        """
+        Generate data for the circuit.
         Args:
-            circuit: The circuit from which to compute the probability.
+            prob_dict(dict): A dictionary whose keys are strings representing the output vectors and
+            their values are the probability to get the output in this circuit.
+            circuit(QuantumCircuit): The circuit that needs to be simulated.
 
         Returns:
-             The probability that the binaomial distribution will use to generate an IQ shot.
+            A dictionary that's filled with the simulated data. The output format is different between
+            measurement level 1 and measurement level 2.
         """
+        # The output is proportional to the number of classical bit.
+        output_length = int(np.sum([creg.size for creg in circuit.cregs]))
+        self._verify_parameters(output_length, prob_dict)
+        prob_arr = self._probability_dict_to_probability_array(prob_dict, output_length)
+        shots = self.options.get("shots")
+        meas_level = self.options.get("meas_level")
+        meas_return = self.options.get("meas_return")
+        run_result = {}
 
-    # pylint: disable=unused-argument
-    def _iq_phase(self, circuit: QuantumCircuit) -> float:
-        """Sub-classes can override this method to introduce a phase in the IQ plan.
+        if meas_level == MeasLevel.CLASSIFIED:
+            counts = {}
+            results = self._rng.multinomial(shots, prob_arr, size=1)[0]
+            for result, num_occurrences in enumerate(results):
+                result_in_str = str(format(result, "b").zfill(output_length))
+                counts[result_in_str] = num_occurrences
+            run_result["counts"] = counts
+        else:
+            phase = self.experiment_helper_object.iq_phase(circuit)
+            memory = self._draw_iq_shots(prob_arr, shots, output_length, phase)
+            if meas_return == "avg":
+                memory = np.average(np.array(memory), axis=0).tolist()
 
-        This is needed, to test the resonator spectroscopy where the point in the IQ
-        plan has a frequency-dependent phase rotation.
-        """
-        return 0.0
+            run_result["memory"] = memory
+        return run_result
 
-    def run(self, run_input, **options):
+    def run(self, run_input: List[QuantumCircuit], **options):
```

Comment:
Need apidoc and type hinting for the return value

Reply:
changed as following:
```python
    def run(self, run_input: List[QuantumCircuit], **options) -> FakeJob:
        """
        Run the IQ backend.
        Args:
            run_input(List[QuantumCircuit]): A list of QuantumCircuit for which the backend will generate
             data for.
            **options: Experiment options. Options that are supported in this backend are 'meas_level'
            and 'meas_return'.
                'meas_level': To generate data in the IQ plain, 'meas_level' should be assigned 1 or
                    MeasLevel.KERNELED. If 'meas_level' is 2 or MeasLevel.CLASSIFIED, the generated data
                    will be in the form of 'counts'.
                'meas_return': This option will only take effect if 'meas_level' = MeasLevel.CLASSIFIED.
                    It can get either MeasReturnType.AVERAGE or MeasReturnType.SINGLE. For the value
                      MeasReturnType.SINGLE the data of each shot will be stored in the result. For
                      MeasReturnType.AVERAGE, an average of all the shots will be calculated and stored
                      in the result.

        Returns:
            FakeJob: A job that contains the simulated data.
        """
```